// concordance/lib/complexValues/dist/global.dev.js
'use strict';

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var constants = require('../constants');

var formatUtils = require('../formatUtils');

var lineBuilder = require('../lineBuilder');

var DEEP_EQUAL = constants.DEEP_EQUAL;
var UNEQUAL = constants.UNEQUAL;

function describe() {
  return new GlobalValue();
}

exports.describe = describe;
exports.deserialize = describe;
var tag = Symbol('GlobalValue');
exports.tag = tag;

var GlobalValue =
/*#__PURE__*/
function () {
  function GlobalValue() {
    _classCallCheck(this, GlobalValue);
  }

  _createClass(GlobalValue, [{
    key: "compare",
    value: function compare(expected) {
      return this.tag === expected.tag ? DEEP_EQUAL : UNEQUAL;
    }
  }, {
    key: "formatDeep",
    value: function formatDeep(theme) {
      return lineBuilder.single(formatUtils.wrap(theme.global, 'Global') + ' ' + theme.object.openBracket + theme.object.closeBracket);
    }
  }]);

  return GlobalValue;
}();

Object.defineProperty(GlobalValue.prototype, 'isComplex', {
  value: true
});
Object.defineProperty(GlobalValue.prototype, 'tag', {
  value: tag
});
