// concordance/lib/complexValues/dist/map.dev.js
'use strict';

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _get(target, property, receiver) { if (typeof Reflect !== "undefined" && Reflect.get) { _get = Reflect.get; } else { _get = function _get(target, property, receiver) { var base = _superPropBase(target, property); if (!base) return; var desc = Object.getOwnPropertyDescriptor(base, property); if (desc.get) { return desc.get.call(receiver); } return desc.value; }; } return _get(target, property, receiver || target); }

function _superPropBase(object, property) { while (!Object.prototype.hasOwnProperty.call(object, property)) { object = _getPrototypeOf(object); if (object === null) break; } return object; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var constants = require('../constants');

var recursorUtils = require('../recursorUtils');

var object = require('./object');

var SHALLOW_EQUAL = constants.SHALLOW_EQUAL;
var UNEQUAL = constants.UNEQUAL;

function describe(props) {
  return new DescribedMapValue(Object.assign({
    size: props.value.size
  }, props));
}

exports.describe = describe;

function deserialize(state, recursor) {
  return new DeserializedMapValue(state, recursor);
}

exports.deserialize = deserialize;
var tag = Symbol('MapValue');
exports.tag = tag;

var MapValue =
/*#__PURE__*/
function (_object$ObjectValue) {
  _inherits(MapValue, _object$ObjectValue);

  function MapValue(props) {
    var _this;

    _classCallCheck(this, MapValue);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(MapValue).call(this, props));
    _this.size = props.size;
    return _this;
  }

  _createClass(MapValue, [{
    key: "compare",
    value: function compare(expected) {
      var result = _get(_getPrototypeOf(MapValue.prototype), "compare", this).call(this, expected);

      if (result !== SHALLOW_EQUAL) return result;
      return this.size === expected.size ? SHALLOW_EQUAL : UNEQUAL;
    }
  }, {
    key: "prepareDiff",
    value: function prepareDiff(expected) {
      // Maps should be compared, even if they have a different number of entries.
      return {
        compareResult: _get(_getPrototypeOf(MapValue.prototype), "compare", this).call(this, expected)
      };
    }
  }, {
    key: "serialize",
    value: function serialize() {
      return [this.size, _get(_getPrototypeOf(MapValue.prototype), "serialize", this).call(this)];
    }
  }]);

  return MapValue;
}(object.ObjectValue);

Object.defineProperty(MapValue.prototype, 'tag', {
  value: tag
});

var DescribedMapValue =
/*#__PURE__*/
function (_object$DescribedMixi) {
  _inherits(DescribedMapValue, _object$DescribedMixi);

  function DescribedMapValue() {
    _classCallCheck(this, DescribedMapValue);

    return _possibleConstructorReturn(this, _getPrototypeOf(DescribedMapValue).apply(this, arguments));
  }

  _createClass(DescribedMapValue, [{
    key: "createIterableRecursor",
    value: function createIterableRecursor() {
      var _this2 = this;

      var size = this.size;
      if (size === 0) return recursorUtils.NOOP_RECURSOR;
      var index = 0;
      var entries;

      var next = function next() {
        if (index === size) return null;

        if (!entries) {
          entries = Array.from(_this2.value);
        }

        var entry = entries[index++];
        return _this2.describeMapEntry(_this2.describeAny(entry[0]), _this2.describeAny(entry[1]));
      };

      return {
        size: size,
        next: next
      };
    }
  }]);

  return DescribedMapValue;
}(object.DescribedMixin(MapValue));

var DeserializedMapValue =
/*#__PURE__*/
function (_object$DeserializedM) {
  _inherits(DeserializedMapValue, _object$DeserializedM);

  function DeserializedMapValue(state, recursor) {
    var _this3;

    _classCallCheck(this, DeserializedMapValue);

    _this3 = _possibleConstructorReturn(this, _getPrototypeOf(DeserializedMapValue).call(this, state[1], recursor));
    _this3.size = state[0];
    return _this3;
  }

  return DeserializedMapValue;
}(object.DeserializedMixin(MapValue));
