// concordance/lib/primitiveValues/dist/bigInt.dev.js
'use strict';

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var constants = require('../constants');

var formatUtils = require('../formatUtils');

var lineBuilder = require('../lineBuilder');

var DEEP_EQUAL = constants.DEEP_EQUAL;
var UNEQUAL = constants.UNEQUAL;

function describe(value) {
  return new BigIntValue(value);
}

exports.describe = describe;
exports.deserialize = describe;
var tag = Symbol('BigIntValue');
exports.tag = tag;

var BigIntValue =
/*#__PURE__*/
function () {
  function BigIntValue(value) {
    _classCallCheck(this, BigIntValue);

    this.value = value;
  }

  _createClass(BigIntValue, [{
    key: "compare",
    value: function compare(expected) {
      return expected.tag === tag && Object.is(this.value, expected.value) ? DEEP_EQUAL : UNEQUAL;
    }
  }, {
    key: "formatDeep",
    value: function formatDeep(theme) {
      return lineBuilder.single(formatUtils.wrap(theme.bigInt, "".concat(this.value, "n")));
    }
  }, {
    key: "serialize",
    value: function serialize() {
      return this.value;
    }
  }]);

  return BigIntValue;
}();

Object.defineProperty(BigIntValue.prototype, 'isPrimitive', {
  value: true
});
Object.defineProperty(BigIntValue.prototype, 'tag', {
  value: tag
});
