// concordance/lib/primitiveValues/dist/boolean.dev.js
'use strict';

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var constants = require('../constants');

var formatUtils = require('../formatUtils');

var lineBuilder = require('../lineBuilder');

var DEEP_EQUAL = constants.DEEP_EQUAL;
var UNEQUAL = constants.UNEQUAL;

function describe(value) {
  return new BooleanValue(value);
}

exports.describe = describe;
exports.deserialize = describe;
var tag = Symbol('BooleanValue');
exports.tag = tag;

var BooleanValue =
/*#__PURE__*/
function () {
  function BooleanValue(value) {
    _classCallCheck(this, BooleanValue);

    this.value = value;
  }

  _createClass(BooleanValue, [{
    key: "compare",
    value: function compare(expected) {
      return this.tag === expected.tag && this.value === expected.value ? DEEP_EQUAL : UNEQUAL;
    }
  }, {
    key: "formatDeep",
    value: function formatDeep(theme) {
      return lineBuilder.single(formatUtils.wrap(theme["boolean"], this.value === true ? 'true' : 'false'));
    }
  }, {
    key: "serialize",
    value: function serialize() {
      return this.value;
    }
  }]);

  return BooleanValue;
}();

Object.defineProperty(BooleanValue.prototype, 'isPrimitive', {
  value: true
});
Object.defineProperty(BooleanValue.prototype, 'tag', {
  value: tag
});
